File: src/computer_build.clj
(ns computer-build
  (:use computer-build.vhdl
        computer-build.state-machine
        clojure.set
        clojure.contrib.pprint))

(defmacro build [cpuname options & instructions]
  `(build* ~cpuname ~options (quote ~instructions)))

(defn rd [port]
  (keyword (str "rd_" (name port))))

(defn wr [port]
  (keyword (str "wr_" (name port))))

(defn binary* [accumulator num]
  (let [last-bit (if (even? num) "0" "1")]
    (cond
      (= 0 num) (str "0" accumulator)
      (= 1 num) (str "1" accumulator)
      true (recur (str last-bit accumulator) (int (/ num 2))))))

(defn binary [width num]
  "Convert number to binary literal format expected in VHDL"
  (let [value (binary* "" num)
        length (count value)]
    (str (apply str (repeat (- width length) "0")) value)))

(defn alu-op-to-opcode [op]
  (if op
    (cond
      (= (name op) "and") "001"
      (= (name op) "complement") "011"
      (= (name op) "+") "100"
      (= (name op) "-") "101"
      (= (name op) "=") "110")
    "000"))

(defn flatten-1 [things]
  (if (empty? things)
    '()
    (if (list? (first things))
      (concat (first things) (flatten-1 (rest things)))
      (cons (first things) (flatten-1 (rest things))))))

(defn mapmap [f m]
  "Replace all values in map m with the result of calling
  f with the value"
  (zipmap (keys m) (map f (vals m))))

(defn rtl-to-microcode [[target _ source & conditional-body]]
  (cond
    (number? source) ; constant-to-register
    {:control-signals (list (wr target)),
     :constant-value source}

    (symbol? source) ; register-to-register
    {:control-signals (list (rd source) (wr target))}

    (= "if" (name target)) ; conditional
    (let [[condition target expectation] _
          body (flatten-1 (map rtl-to-microcode (cons source conditional-body)))]
      (list
        ; load target
        {:control-signals (list (rd target) (wr :alu_a)) }
        ; load expectation and compare
        (if (number? expectation)
          {:control-signals (list (wr :alu_b))
           :constant-value expectation
           :alu_op condition
           :conditional true
           :body body}
          {:control-signals (list (rd expectation) (wr :alu_b))
           :alu_op condition
           :conditional true
           :body body})
        ))

    (and (seq? source) (symbol? (first source))) ; ALU-to-register
    (let [[alu_op operand_a operand_b] source]
      (list
        {:control-signals (list (rd operand_a) (wr :alu_a)) :alu_op alu_op}
        (if (= 3 (count source))
          (if (number? operand_b)
            {:control-signals (list (wr :alu_b)) :alu_op alu_op
              :constant-value operand_b}
            {:control-signals (list (rd operand_b) (wr :alu_b)) :alu_op alu_op}))
        {:control-signals (list :rd_alu (wr target)) :alu_op alu_op}))))

(defn name-for-state [instruction-name index]
  (keyword (str instruction-name "_" index)))

(defn link-state [instruction-name last-index body index]
  (let [next-state
          (if (= index last-index)
            :fetch
            (name-for-state instruction-name (+ index 1)))
        connected-body
          {(name-for-state instruction-name index)
            (assoc body :next next-state)}]
  (if-let [conditional-body (:body body)]
    ; conditional
    (let [instruction-name (str instruction-name "_" index)
          last-index (dec (count conditional-body))]
      (merge connected-body
             (apply merge (map (partial link-state instruction-name last-index)
                               conditional-body (iterate inc 0)))))
    ; not conditional
    connected-body)))

(defn make-states-for-instruction [[_ instruction-name & RTLs]]
  (let [microcode (flatten-1 (map rtl-to-microcode RTLs))
        last-index (- (count microcode) 1)]
    (apply merge (map (partial link-state instruction-name last-index)
                      microcode (iterate inc 0)))))

(defn make-states [instructions]
  "Given a set of instructions, create the set of states
  necessary to execute their microcode"
  (apply merge (map make-states-for-instruction instructions)))

(defn make-opcodes [instructions]
  "Given a set of instructions, assign opcodes to each one in
  a map with keys being instruction names"
  (let [names (map #(nth % 1) instructions)
        width (Math/ceil (/ (Math/log (count names)) (Math/log 2)))
        to-binary (fn [n] (let [raw (Integer/toString n 2)
                                padlength (- width (count raw))]
                            (str (apply str (repeat padlength "0")) raw)))
    op-values (map to-binary (range (count names)))]
    (zipmap names op-values)))

(defn realize-state [control-signals state]
  (let [highs (:control-signals state)
        assertions (map #(list 'high %) highs)
        clears (map #(list 'low %) (difference control-signals highs))]
    (vec (concat
           (:code state)
           assertions
           clears
           (if-let [const (:constant-value state)]
             `((<= :system_bus ~(binary 8 const)))
             `((<= :system_bus ~(apply str (repeat 8 "Z")))))
           `((<= :alu_operation ~(alu-op-to-opcode (:alu_op state))))))))

(defn control-unit [instructions]
  "Given a set of states, make a control unit that will
  execute them"
  (let [opcodes (make-opcodes instructions)
        opcode-width (count (second (first opcodes)))
        static-states {:fetch 
                         {:control-signals '(:rd_pc, :wr_MA), :next :store_instruction}
                       :store_instruction
                         {:next :decode,
                          :control-signals '(:rd_MD, :wr_IR, :inc_pc),
                          :code
                          [`(if (and (event :clock) (= :clock 0))
                              (<= :opcode ~(- opcode-width 1) 0
                                  :system_bus 7 ~(- 8 opcode-width)))]}
                       :decode {:control-signals '()}}
        states (merge (make-states instructions) static-states)
        conditional-states (select-keys states (for [[k v] states :when (:conditional v)] k))
        unconditional-states (select-keys states (for [[k v] states :when (not (:conditional v))] k))
        control-signals (set (apply concat (map
                                             (fn [[_ body]] (:control-signals body))
                                             states)))
        inputs {:reset std-logic, :condition std-logic}
        outputs (assoc
                  (zipmap control-signals (repeat (count control-signals) std-logic))
                  :alu_operation (std-logic-vector 2 0))]

    (list (state-machine "control_unit"
               ; inputs
               inputs
               ; outputs
               outputs
               ; input/outputs
               {:system_bus (std-logic-vector 7 0)}
               ; signals
               { :opcode (std-logic-vector (- opcode-width 1) 0) }
               ; reset
               (list* '(<= :alu_operation "000")
                      '(goto :fetch)
                      '(<= :system_bus "ZZZZZZZZ")
                      (map #(list 'low %) control-signals))
               ; states
               (mapmap (partial realize-state control-signals) states)
               ; transitions
               (concat
                 ; states
                 (map (fn [[k v]] (list k (:next v))) (dissoc unconditional-states :decode))
                 ; conditional false
                 (map (fn [[k v]] (list k '(= :condition 0) (:next v))) conditional-states)
                 ; conditional true
                 (map (fn [[k v]] (list k '(= :condition 1) (name-for-state (name k) 0))) conditional-states)

                 ; decode
                 (map #(list
                         ; from
                         :decode
                         ; condition
                         `(= :opcode ~(second %))
                         ; to
                         (keyword (str (first %) "_0")))
                      ; for each opcode
                      opcodes)))
          (assoc inputs :clock std-logic) outputs)))

(defn build* [cpuname options instructions]
  (.mkdir (java.io.File. cpuname))
  (let [[control-unit control-in control-out] (control-unit instructions)
        control-signals (concat (dissoc control-in :clock :reset) control-out)
        dynamic-signals (map (fn [[k v]] (list 'signal k v)) control-signals)]
    (with-open [main-vhdl (java.io.FileWriter. (str cpuname "/main.vhdl"))
                control-vhdl (java.io.FileWriter. (str cpuname "/control.vhdl"))]
      (pprint dynamic-signals)
      (binding [*out* control-vhdl]
        (generate-vhdl control-unit))
      (binding [*out* main-vhdl]
        (generate-vhdl `(entity "main"
          ; ports
          [(:clock :in ~std-logic)
           (:reset :in ~std-logic)
           (:bus_inspection :out ~(std-logic-vector 7 0))]
          ; defs
          [~@dynamic-signals
           (signal :system_bus ~(std-logic-vector 7 0))
                    
          (component :reg
            (:clock :in ~std-logic)
            (:data_in :in ~(std-logic-vector 7 0))
            (:data_out :out ~(std-logic-vector 7 0))
            (:wr :in ~std-logic)
            (:rd :in ~std-logic))

          (component :program_counter
            (:clock :in ~std-logic)
            (:data_in :in ~(std-logic-vector 7 0))
            (:data_out :out ~(std-logic-vector 7 0))
            (:wr :in ~std-logic)
            (:rd :in ~std-logic)
            (:inc :in ~std-logic))

          (component :ram
            (:clock :in ~std-logic)
            (:data_in :in ~(std-logic-vector 7 0))
            (:data_out :out ~(std-logic-vector 7 0))
            (:address :in ~(std-logic-vector 4 0))
            (:wr_data :in ~std-logic)
            (:wr_addr :in ~std-logic)
            (:rd :in ~std-logic))

          (component :alu
            (:clock :in ~std-logic)
            (:data_in :in ~(std-logic-vector 7 0))
            (:data_out :out ~(std-logic-vector 7 0))
            (:op :in ~(std-logic-vector 2 0))
            (:wr_a :in ~std-logic)
            (:wr_b :in ~std-logic)
            (:rd :in ~std-logic)
            (:condition :out ~std-logic))

          (component :control_unit
            ~@(concat (map input control-in)
                      (map output control-out)
                      `((:system_bus :inout ~(std-logic-vector 7 0)))))]
          ; architecture
          [
          (instance :program_counter "pc" :clock :system_bus :system_bus
                    :wr_pc :rd_pc, :inc_pc)
          ; instruction register
          (instance :reg "ir" :clock :system_bus :system_bus :wr_IR :rd_IR)
          ; accumulator
          (instance :reg "A" :clock :system_bus :system_bus :wr_A :rd_A)
          (instance :ram "main_memory" :clock :system_bus :system_bus
                    ~(subbits :system_bus 4 0) :wr_MD :wr_MA :rd_MD)
          (instance :alu "alu0" :clock :system_bus :system_bus :alu_operation
                    :wr_alu_a :wr_alu_b :rd_alu, :condition)
          (instance :control_unit "control0"
                    ; same ports as the control signals we got
                    ~@(map first (concat control-in control-out)) :system_bus)
          (<= :bus_inspection :system_bus)
          ]))))))

File: lib/Computer.Build.rb
require 'computer_build/vhdl'
require 'computer_build/state_machine'

class Computer
  include VHDL::Helpers

  def self.Build(name)
    instance = Computer.new(name)
    yield(instance)
    instance.generate
  end
  
  attr_writer :address_width
  
  def initialize(name)
    @name = name
    @instructions = []
  end
  
  # DSL method
  def instruction(name)
    inst = Instruction.new(name)
    yield(inst)
    @instructions << inst
  end
  
  def generate
    states = make_states(@instructions).merge(static_states)
    opcodes = make_opcodes(@instructions)
    opcode_length = opcodes.values.first.length
    control_signals = states.values.map(&:control_signals).flatten.uniq

    control = state_machine("control_unit") do |m|
      m.input :reset, VHDL::STD_LOGIC
      m.input :condition, VHDL::STD_LOGIC
      m.inout :system_bus, VHDL::STD_LOGIC_VECTOR(7..0)
      m.output :alu_operation, VHDL::STD_LOGIC_VECTOR(2..0)
      control_signals.each do |sig|
        m.output sig, VHDL::STD_LOGIC
      end

      m.signal :opcode,
        VHDL::STD_LOGIC_VECTOR((opcode_length-1)..0)

      states.values.map(&:constant_value).compact.uniq.each do |const|
        m.constant "CONSTANT_#{const}", VHDL::STD_LOGIC_VECTOR(7..0), const.to_logic(8)
      end

      m.reset do |r|
        r.goto :fetch
        control_signals.each do |sig|
          r.low sig.to_sym
        end
        r.assign :alu_operation, "000"
        r.assign :system_bus, "ZZZZZZZZ"
      end

      states.each do |name, state|
        m.state(name) do |s|
          control_signals.each do |sig|
            s.assign sig, state.control_signals.include?(sig) ? '1' : '0'
          end
          s.assign :alu_operation, state.alu_op ? state.alu_op.opcode : "000"

          if state.constant_value
            s.assign :system_bus, "CONSTANT_#{state.constant_value}".to_sym
          else
            s.assign :system_bus, "ZZZZZZZZ"
          end

          if name == 'store_instruction'
            s.if event(:clock), VHDL::Equal.new(:clock,"0") do |thn|
              thn.assign :opcode, "#{opcode_length-1} downto 0",
                :system_bus, "7 downto #{7-opcode_length+1}"
            end
          end
        end

        if state.condition
          m.transition :from => name, :to => state.next,
            :on => VHDL::Equal.new(:condition, "0")
          m.transition :from => name, :to => name+"_0",
            :on => VHDL::Equal.new(:condition, "1")
        else
          m.transition :from => name, :to => state.next if state.next
        end
      end

      # instruction decode
      opcodes.each do |instruction, opcode|
        m.transition :from => :decode, :to => instruction.name+"_0",
          :on => equal(:opcode, opcode)
      end
    end

    main = entity("main") do |e|
      e.port :clock ,:in, VHDL::STD_LOGIC
      e.port :reset, :in, VHDL::STD_LOGIC
      e.port :bus_inspection, :out, VHDL::STD_LOGIC_VECTOR(7..0)

      e.signal :system_bus, VHDL::STD_LOGIC_VECTOR(7..0)
      e.signal :alu_operation, VHDL::STD_LOGIC_VECTOR(2..0)
      e.signal :alu_condition, VHDL::STD_LOGIC

      control_signals.each do |sig|
        e.signal sig, VHDL::STD_LOGIC
      end

      e.component :reg do |c|
        c.in :clock, VHDL::STD_LOGIC
        c.in :data_in, VHDL::STD_LOGIC_VECTOR(7..0)
        c.out :data_out, VHDL::STD_LOGIC_VECTOR(7..0)
        c.in :wr, VHDL::STD_LOGIC
        c.in :rd, VHDL::STD_LOGIC
      end

      e.component :program_counter do |c|
        c.in :clock, VHDL::STD_LOGIC
        c.in :data_in, VHDL::STD_LOGIC_VECTOR(7..0)
        c.out :data_out, VHDL::STD_LOGIC_VECTOR(7..0)
        c.in :wr, VHDL::STD_LOGIC
        c.in :rd, VHDL::STD_LOGIC
        c.in :inc, VHDL::STD_LOGIC
      end

      e.component :ram do |c|
        c.in :clock, VHDL::STD_LOGIC
        c.in :data_in, VHDL::STD_LOGIC_VECTOR(7..0)
        c.out :data_out, VHDL::STD_LOGIC_VECTOR(7..0)
        c.in :address, VHDL::STD_LOGIC_VECTOR(4..0)
        c.in :wr_data, VHDL::STD_LOGIC
        c.in :wr_addr, VHDL::STD_LOGIC
        c.in :rd, VHDL::STD_LOGIC
      end

      e.component :alu do |c|
        c.in :clock, VHDL::STD_LOGIC
        c.in :data_in, VHDL::STD_LOGIC_VECTOR(7..0)
        c.out :data_out, VHDL::STD_LOGIC_VECTOR(7..0)
        c.in :op, VHDL::STD_LOGIC_VECTOR(2..0)
        c.in :wr_a, VHDL::STD_LOGIC
        c.in :wr_b, VHDL::STD_LOGIC
        c.in :rd, VHDL::STD_LOGIC
        c.out :condition, VHDL::STD_LOGIC
      end

      e.component :control_unit do |c|
        c.in :clock, VHDL::STD_LOGIC
        c.in :reset, VHDL::STD_LOGIC
        c.in :condition, VHDL::STD_LOGIC
        c.out :alu_operation, VHDL::STD_LOGIC_VECTOR(2..0)
        control_signals.each do |sig|
          c.out sig, VHDL::STD_LOGIC
        end
        c.inout :system_bus, VHDL::STD_LOGIC_VECTOR(7..0)
      end

      e.behavior do |b|
        b.instance :program_counter, "pc", :clock, :system_bus, :system_bus,
          :wr_pc, :rd_pc, :inc_pc
        b.instance :reg, "ir", :clock, :system_bus, :system_bus, :wr_IR, :rd_IR
        b.instance :reg, "A", :clock, :system_bus, :system_bus, :wr_A, :rd_A
        b.instance :ram, "main_memory", :clock, :system_bus, :system_bus,
          subbits(:system_bus, 4..0), :wr_MD, :wr_MA, :rd_MD
        b.instance :alu, "alu0", :clock, :system_bus, :system_bus,
          :alu_operation, :wr_alu_a, :wr_alu_b, :rd_alu, :alu_condition
        b.instance :control_unit, "control0", [:clock, :reset, :alu_condition,
            :alu_operation] + control_signals + [:system_bus]
        b.assign :bus_inspection, :system_bus
      end
    end
    
    Dir.mkdir @name rescue nil # ignore error if dir already exists
    File.open(File.join(@name, 'control.vhdl'), 'w') do |f|
      generate_vhdl(control, f)
    end

    File.open(File.join(@name, 'main.vhdl'), 'w') do |f|
      generate_vhdl(main, f)
    end
  end
  
  # inner classes
  
  class Instruction
    attr_reader :name, :steps

    def initialize(name)
      @name = name
      @steps = []
    end
    
    def move(target, source)
      @steps << RTL.new(target, source)
    end

    def microcode
      steps.map(&:to_microcode).flatten
    end

    def if(condition, &body)
      @steps << Conditional.new(condition, &body)
    end
  end
  
  class ALUOperation
    attr_reader :op, :operands

    def initialize(op, *operands)
      @op = op
      @operands = operands
    end

    def opcode
      return {
        :and        => "001",
        :or         => "010",
        :complement => "011",
        :add        => "100",
        :subtract   => "101",
        :equal      => "110",
        :lessthan   => "111"}[@op]
    end
  end

  private

  class Conditional
    def initialize(condition)
      @condition = condition # instance of ALUOperation
      @steps = []
      @true_body = []
      yield self
    end

    # DSL method
    def move(target, source)
      @steps << RTL.new(target, source)
    end

    def to_microcode
      steps = []
      steps << MicrocodeState.new do |state|
        state.control_signals = ["rd_#{@condition.operands.first}", "wr_alu_a"]
        state.alu_op = @condition
      end
      conditional = MicrocodeState.new do |state|
        op = @condition.operands.last
        state.control_signals = ["wr_alu_b"]

        if op.is_a? Fixnum
          state.constant_value = op
        else
          state.control_signals += "rd_#{op}"
        end

        state.alu_op = @condition
        state.condition = @condition
      end
      steps << conditional

      body = @steps.map(&:to_microcode).flatten
      conditional.body_size = body.length

      body.each {|state| state.conditional = conditional}

      return steps + body
    end
  end

  class RTL
    def initialize(target, source)
      @target = target
      @source = source
    end
    
    def to_microcode
      if @source.is_a? Fixnum
        return MicrocodeState.new do |state|
          state.control_signals = "wr_#{@target}"
          state.constant_value = @source
        end
      elsif @source.is_a? Symbol
        return MicrocodeState.new do |state|
          state.control_signals = ["wr_#{@target}", "rd_#{@source}"]
        end
      elsif @source.is_a? ALUOperation
        steps = []
        steps << MicrocodeState.new do |state|
          state.control_signals = ["rd_#{@source.operands.first}", "wr_alu_a"]
          state.alu_op = @source
        end

        if @source.operands.length == 2
          steps << MicrocodeState.new do |state|
            state.control_signals = ["wr_alu_b"]
            if @source.operands.last.is_a? Fixnum
              state.constant_value = @source.operands.last
            else
              state.control_signals << "rd_#{@source.operands.last}"
            end
          end
        end

        steps << MicrocodeState.new do |state|
          state.control_signals = ["rd_alu", "wr_#{@target}"]
          state.alu_op = @source
        end
        return steps
      end
    end
  end

  class MicrocodeState
    attr_accessor :control_signals, :alu_op, :constant_value, :next,
      :condition, :conditional, :body_size, :index

    def initialize(&blk)
      yield(self) if blk
    end
  end

  def make_states(instructions)
    states = {}
    instructions.each do |instr|
      steps = instr.microcode
      indexes = {nil => 0}
      steps.each do |step|
        indexes[step.conditional] ||= 0
        index = indexes[step.conditional]

        if step.conditional
          if index < step.conditional.body_size - 1
            step.next = instr.name+"_"+(step.conditional.index.to_s)+"_"+(index+1).to_s
          else
            step.next = step.conditional.next
          end
          states[instr.name+"_"+(step.conditional.index.to_s)+"_"+index.to_s] = step
        else
          if index < steps.reject(&:conditional).length - 1
            step.next = instr.name+"_"+(index+1).to_s
          else
            step.next = :fetch
          end
          states[instr.name+"_"+index.to_s] = step
        end
        step.index = index
        indexes[step.conditional] += 1
      end
    end

    return states
  end

  def make_opcodes(instructions)
    bits = (Math.log(instructions.length)/Math.log(2)).ceil
    opcodes = {}
    instructions.each_with_index do |instruction, idx|
      bin_string = idx.to_s(2)
      bin_string = ("0" * (bits - bin_string.length)) + bin_string
      opcodes[instruction] = bin_string
    end
    return opcodes
  end

  def static_states
    states = {}
    states['fetch'] = MicrocodeState.new do |s|
      s.control_signals = ['rd_pc', 'wr_MA']
      s.next = 'store_instruction'
    end

    states['store_instruction'] = MicrocodeState.new do |s|
      s.control_signals = ['rd_MD', 'wr_IR', 'inc_pc']
      s.next = 'decode'
    end

    states['decode'] = MicrocodeState.new do |s|
      s.control_signals = []
    end

    return states
  end
end

def complement(value)
  Computer::ALUOperation.new(:complement, value)
end

def add(operand1, operand2)
  Computer::ALUOperation.new(:add, operand1, operand2)
end

def bitwise_and(operand1, operand2)
  Computer::ALUOperation.new(:and, operand1, operand2)
end

def subtract(operand1, operand2)
  Computer::ALUOperation.new(:subtract, operand1, operand2)
end

def equal(operand1, operand2)
  Computer::ALUOperation.new(:equal, operand1, operand2)
end
